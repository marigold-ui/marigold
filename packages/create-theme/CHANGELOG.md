# @marigold/create-theme

## 2.0.2

### Patch Changes

- Updated dependencies [[`32353f56`](https://github.com/marigold-ui/marigold/commit/32353f56728ddc02e8b4ec848184b7dc5e5f9d19)]:
  - @marigold/tokens@3.1.0

## 2.0.1

### Patch Changes

- Updated dependencies [[`a41bb8a3`](https://github.com/marigold-ui/marigold/commit/a41bb8a39ace6006bfc2351f6d4649305bc1d286)]:
  - @marigold/tokens@3.0.0

## 2.0.0

### Major Changes

- [#1891](https://github.com/marigold-ui/marigold/pull/1891) [`74c91e28`](https://github.com/marigold-ui/marigold/commit/74c91e289ffadfc082a548fc55ab9bd315ac1d8a) Thanks [@sebald](https://github.com/sebald)! - feat(tokens): Separate size and space scale

### Patch Changes

- Updated dependencies [[`74c91e28`](https://github.com/marigold-ui/marigold/commit/74c91e289ffadfc082a548fc55ab9bd315ac1d8a)]:
  - @marigold/tokens@2.0.0

## 1.0.0

### Major Changes

- [#1883](https://github.com/marigold-ui/marigold/pull/1883) [`af566de4`](https://github.com/marigold-ui/marigold/commit/af566de483e80bfedcd09b073be8559eebf17fca) Thanks [@sebald](https://github.com/sebald)! - feat: rename scales

### Patch Changes

- Updated dependencies [[`d436930f`](https://github.com/marigold-ui/marigold/commit/d436930f7d88b572c4365a442a0914bba1147657), [`af566de4`](https://github.com/marigold-ui/marigold/commit/af566de483e80bfedcd09b073be8559eebf17fca), [`d2c3eb50`](https://github.com/marigold-ui/marigold/commit/d2c3eb50ceb9941926e7e658dcba7a74cb87e6d7), [`ec4f8f22`](https://github.com/marigold-ui/marigold/commit/ec4f8f228f05e88f2b9f17784b59f3cfc8ceabe7)]:
  - @marigold/components@0.7.0
  - @marigold/tokens@1.0.0
  - @marigold/system@0.7.0
