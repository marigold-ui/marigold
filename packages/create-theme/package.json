{
  "name": "@marigold/create-theme",
  "version": "2.0.2",
  "description": "Marigold helper to kickoff Themes",
  "license": "MIT",
  "keywords": [
    "marigold",
    "theme",
    "styles"
  ],
  "main": "./dist/index.js",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "sideEffects": false,
  "files": [
    "dist"
  ],
  "repository": {
    "type": "git",
    "url": "github:marigold-ui/marigold",
    "directory": "themes/create-theme"
  },
  "dependencies": {
    "@marigold/tokens": "workspace:*"
  },
  "devDependencies": {
    "@marigold/components": "workspace:*",
    "@marigold/system": "workspace:*",
    "@marigold/tsconfig": "workspace:*",
    "tsup": "6.4.0"
  },
  "scripts": {
    "build": "tsup src/index.ts",
    "watch": "tsup src/index.ts --watch",
    "clean": "rm -rf node_modules && rm -rf dist"
  }
}
