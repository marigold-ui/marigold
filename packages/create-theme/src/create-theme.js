import * as Token from '@marigold/tokens';
export const base = {
  breakpoints: ['40em', '52em', '64em'],
  fontWeights: Token.typography.weight,
  lineHeights: Token.typography.lineHeight,
  letterSpacings: Token.typography.letterSpacing,
  borderWidths: Token.border.width,
  radii: Token.border.radius,
  zIndices: Token.layer,
  opacities: Token.opacity,
  components: {},
};
/*#__PURE__*/
export const createTheme = ({
  colors: configColors = [],
  fonts: configFonts,
  typography = 'fixed',
  dimensions = 'fixed',
  components = {},
}) => {
  // Create colors (pick from Token or custom)
  let colors = {};
  for (let i = 0; i < configColors.length; i++) {
    const color = configColors[i];
    if (typeof color === 'string') {
      colors[color] = Token.color[color];
    } else {
      colors[color.name] = color.value;
    }
  }
  return {
    // Add base scales
    ...base,
    // Add selected colors
    colors,
    // Use fixed or fluid dimensions
    space: Token.space[dimensions],
    sizes: Token.size[dimensions],
    // Use fallback font if none provided
    fonts: configFonts || Token.typography.font,
    // Use fixed or fluid typography
    fontSizes: Token.typography.size[typography],
    // Add component styles
    components,
  };
};
//# sourceMappingURL=create-theme.js.map
