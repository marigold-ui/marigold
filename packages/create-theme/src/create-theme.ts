import type { ScaleValue, Theme as Scales } from '@marigold/system';
import type { Theme } from '@marigold/components';

import * as Token from '@marigold/tokens';

export const base: Theme = {
  breakpoints: ['40em', '52em', '64em'],

  fontWeights: Token.typography.weight,
  lineHeights: Token.typography.lineHeight,
  letterSpacings: Token.typography.letterSpacing,

  borderWidths: Token.border.width,
  radii: Token.border.radius,

  zIndices: Token.layer,
  opacities: Token.opacity,
};

type AvailableColor = keyof typeof Token.color;
type CustomColor = { name: string; value: ScaleValue<string> };

export interface ThemeConfig {
  /**
   * Used to pick from existing color palette or define custom colors.
   */
  colors?: (AvailableColor | CustomColor)[];

  /**
   * Used to define a the `font-family` scale.
   */
  fonts?: { [key: string]: string };

  /**
   * Used to define if fixed or fluid typography should be used.
   */
  typography?: 'fixed' | 'fluid';

  /**
   * Used to define if fluid or fixed sizes and whitespace should be used.
   */
  dimensions?: 'fixed' | 'fluid';

  components?: Omit<Theme, keyof Scales>;
}

/*#__PURE__*/
export const createTheme = ({
  colors: configColors = [],
  fonts: configFonts,
  typography = 'fixed',
  dimensions = 'fixed',
  components,
}: ThemeConfig): Theme => {
  // Create colors (pick from Token or custom)
  let colors: Theme['colors'] = {};
  for (let i = 0; i < configColors.length; i++) {
    const color = configColors[i];
    if (typeof color === 'string') {
      colors[color] = Token.color[color];
    } else {
      colors[color.name] = color.value;
    }
  }

  return {
    // Add base scales
    ...base,

    // Add selected colors
    colors,

    // Use fixed or fluid dimensions
    space: Token.space[dimensions],
    sizes: Token.size[dimensions],

    // Use fallback font if none provided
    fonts: configFonts || Token.typography.font,

    // Use fixed or fluid typography
    fontSizes: Token.typography.size[typography],

    // Add component styles
    ...components,
  };
};
