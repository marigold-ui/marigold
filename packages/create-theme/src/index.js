export * from './create-theme';
//# sourceMappingURL=index.js.map
