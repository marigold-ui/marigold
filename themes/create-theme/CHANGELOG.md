# @marigold/create-theme

## 1.0.0

### Patch Changes

- [#1809](https://github.com/marigold-ui/marigold/pull/1809) [`539d4198`](https://github.com/marigold-ui/marigold/commit/539d41987118db125ee37a4a83231335ea15830a) Thanks [@sebald](https://github.com/sebald)! - feat(create-theme): Introduce helper package to create themes with sensible defaults

- Updated dependencies [[`4cc0ad3b`](https://github.com/marigold-ui/marigold/commit/4cc0ad3b85b993e01b4d85b6f30b1a81cdee2351), [`d162a6a8`](https://github.com/marigold-ui/marigold/commit/d162a6a8d1f9d99bfc08d59f6dae294a6fa95310), [`539d4198`](https://github.com/marigold-ui/marigold/commit/539d41987118db125ee37a4a83231335ea15830a), [`e053b7b9`](https://github.com/marigold-ui/marigold/commit/e053b7b903f02c56cf10e6b9aecbedd29399895d), [`527ba94f`](https://github.com/marigold-ui/marigold/commit/527ba94fa1a3255dc6f846fcc9def978ec906bf3), [`b55e10bc`](https://github.com/marigold-ui/marigold/commit/b55e10bc7119554373d43f9d8872346c1e6d6c39), [`9a95a6f5`](https://github.com/marigold-ui/marigold/commit/9a95a6f541dec34e27eac8557186f929e811c07c), [`f62033b6`](https://github.com/marigold-ui/marigold/commit/f62033b642a623a75f311af701febc4b54f77120), [`21f72b34`](https://github.com/marigold-ui/marigold/commit/21f72b3499c7ed2f88a0b96315f0374140e98e85), [`25c8675e`](https://github.com/marigold-ui/marigold/commit/25c8675ed470cc71f8f968a83b9781b66412c493)]:
  - @marigold/system@0.6.0
  - @marigold/components@0.6.0
