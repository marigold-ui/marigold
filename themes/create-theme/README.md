# `@marigold/create-theme`

> Helper to create new themes for `@marigold/components`
