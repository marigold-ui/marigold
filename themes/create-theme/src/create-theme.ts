import type { ScaleValue, Theme as Scales } from '@marigold/system';
import type { Theme } from '@marigold/components';

import * as Vars from './vars';

export const base: Theme = {
  breakpoints: ['40em', '52em', '64em'],

  fontWeights: Vars.typography.weight,
  lineHeights: Vars.typography.lineHeight,
  letterSpacings: Vars.typography.letterSpacing,

  borderWidths: Vars.border.width,
  radii: Vars.border.radius,

  zIndices: Vars.layer,
  opacities: Vars.opacity,
};

type AvailableColor = keyof typeof Vars.color;
type CustomColor = { name: string; value: ScaleValue<string> };

export interface ThemeConfig {
  /**
   * Used to pick from existing color palette or define custom colors.
   */
  colors?: (AvailableColor | CustomColor)[];

  /**
   * Used to define a the `font-family` scale.
   */
  fonts?: { [key: string]: string };

  /**
   * Used to define if fixed or fluid typography should be used.
   */
  typography?: 'fixed' | 'fluid';

  /**
   * Used to define if fluid or fixed sizes and whitespace should be used.
   */
  dimensions?: 'fixed' | 'fluid';

  components?: Omit<Theme, keyof Scales>;
}

/*#__PURE__*/
export const createTheme = ({
  colors: configColors = [],
  fonts: configFonts,
  typography = 'fixed',
  dimensions = 'fixed',
  components,
}: ThemeConfig): Theme => {
  // Create colors (pick from vars or custom)
  let colors: Theme['colors'] = {};
  for (let i = 0; i < configColors.length; i++) {
    const color = configColors[i];
    if (typeof color === 'string') {
      colors[color] = Vars.color[color];
    } else {
      colors[color.name] = color.value;
    }
  }

  return {
    // Add base scales
    ...base,

    // Add selected colors
    colors,

    // Use fixed or fluid dimensions
    space: Vars.size[dimensions],
    sizes: Vars.size[dimensions],

    // Use fallback font if none provided
    fonts: configFonts || Vars.typography.font,

    // Use fixed or fluid typography
    fontSizes: Vars.typography.size[typography],

    // Add component styles
    ...components,
  };
};
