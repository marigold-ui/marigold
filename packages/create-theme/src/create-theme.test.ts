import { createTheme } from './create-theme';

test.each([
  ['breakpoints'],
  ['fontWeights'],
  ['lineHeights'],
  ['letterSpacings'],
  ['borderWidths'],
  ['radii'],
  ['zIndices'],
  ['opacities'],
])('created theme comes with defaults for "%s"', args => {
  const theme = createTheme({});
  expect(theme).toHaveProperty(args);
  // @ts-ignore
  expect(theme[args]).toMatchSnapshot();
});

test('allows to pick colors', () => {
  expect(createTheme({ colors: [] }).colors).toMatchInlineSnapshot(`{}`);
  expect(createTheme({ colors: ['brand'] }).colors).toMatchInlineSnapshot(`
    {
      "brand": {
        "primary": "#fa8005",
      },
    }
  `);
  expect(createTheme({ colors: ['red', 'blue', 'violet'] }).colors)
    .toMatchInlineSnapshot(`
    {
      "blue": {
        "00": "#e7f5ff",
        "10": "#d0ebff",
        "20": "#a5d8ff",
        "30": "#74c0fc",
        "40": "#4dabf7",
        "50": "#339af0",
        "60": "#228be6",
        "70": "#1c7ed6",
        "80": "#1971c2",
        "90": "#1864ab",
      },
      "red": {
        "00": "#fff5f5",
        "10": "#ffe3e3",
        "20": "#ffc9c9",
        "30": "#ffa8a8",
        "40": "#ff8787",
        "50": "#ff6b6b",
        "60": "#fa5252",
        "70": "#f03e3e",
        "80": "#e03131",
        "90": "#c92a2a",
      },
      "violet": {
        "00": "#f3f0ff",
        "10": "#e5dbff",
        "20": "#d0bfff",
        "30": "#b197fc",
        "40": "#9775fa",
        "50": "#845ef7",
        "60": "#7950f2",
        "70": "#7048e8",
        "80": "#6741d9",
        "90": "#5f3dc4",
      },
    }
  `);
});

test('allows to add custom colors', () => {
  expect(
    createTheme({
      colors: [
        'violet',
        { name: 'red', value: '#ff0000' },
        { name: 'blue', value: '#0000ff' },
        {
          name: 'unicorn',
          value: {
            one: '#fffafa',
            two: '#fff7b7',
            three: '#ffc6fb',
            four: '#d6b9ff',
            five: '#c188d7',
          },
        },
      ],
    }).colors
  ).toMatchInlineSnapshot(`
    {
      "blue": "#0000ff",
      "red": "#ff0000",
      "unicorn": {
        "five": "#c188d7",
        "four": "#d6b9ff",
        "one": "#fffafa",
        "three": "#ffc6fb",
        "two": "#fff7b7",
      },
      "violet": {
        "00": "#f3f0ff",
        "10": "#e5dbff",
        "20": "#d0bfff",
        "30": "#b197fc",
        "40": "#9775fa",
        "50": "#845ef7",
        "60": "#7950f2",
        "70": "#7048e8",
        "80": "#6741d9",
        "90": "#5f3dc4",
      },
    }
  `);
});

test('allows to set fonts', () => {
  expect(
    createTheme({
      fonts: { body: 'Arial', heading: 'Gothic' },
    }).fonts
  ).toMatchInlineSnapshot(`
    {
      "body": "Arial",
      "heading": "Gothic",
    }
  `);
});

test('uses fallback fonts if none provided', () => {
  expect(createTheme({}).fonts).toMatchInlineSnapshot(`
    {
      "mono": "Dank Mono,Operator Mono,Inconsolata,Fira Mono,ui-monospace,SF Mono,Monaco,Droid Sans Mono,Source Code Pro,monospace",
      "sans": "system-ui,-apple-system,Segoe UI,Roboto,Ubuntu,Cantarell,Noto Sans,sans-serif",
      "serif": "ui-serif,serif",
    }
  `);
});

test('uses fixed values for typography by default', () => {
  expect(createTheme({}).fontSizes).toMatchInlineSnapshot(`
    {
      "large-1": "2.5rem",
      "large-2": "3rem",
      "large-3": "3.5rem",
      "medium-1": "1.1rem",
      "medium-2": "1.25rem",
      "medium-3": "1.5rem",
      "medium-4": "2rem",
      "small-1": ".5rem",
      "small-2": ".75rem",
      "small-3": "1rem",
    }
  `);

  expect(createTheme({}).fontSizes).toEqual(
    createTheme({ typography: 'fixed' }).fontSizes
  );
});

test('allow to use fluid typography', () => {
  expect(createTheme({ typography: 'fluid' }).fontSizes).toMatchInlineSnapshot(`
    {
      "large-1": "clamp(2rem, 9vw, 3.5rem)",
      "large-2": "clamp(3.5rem, 12vw, 5rem)",
      "medium-1": "clamp(1rem, 4vw, 1.5rem)",
      "medium-2": "clamp(1.5rem, 6vw, 2.5rem)",
      "small-1": "clamp(.75rem, 2vw, 1rem)",
    }
  `);
});

test('uses fixed values for dimensaions by default', () => {
  expect(createTheme({}).space).toMatchInlineSnapshot(`
    {
      "large-1": "2rem",
      "large-2": "3rem",
      "large-3": "4rem",
      "large-4": "5rem",
      "medium-1": "1.25rem",
      "medium-2": "1.5rem",
      "medium-3": "1.75rem",
      "none": 0,
      "small-1": ".25rem",
      "small-2": ".5rem",
      "small-3": "1rem",
    }
  `);
  expect(createTheme({}).space).toMatchInlineSnapshot(`
    {
      "large-1": "2rem",
      "large-2": "3rem",
      "large-3": "4rem",
      "large-4": "5rem",
      "medium-1": "1.25rem",
      "medium-2": "1.5rem",
      "medium-3": "1.75rem",
      "none": 0,
      "small-1": ".25rem",
      "small-2": ".5rem",
      "small-3": "1rem",
    }
  `);

  expect(createTheme({}).space).toEqual(
    createTheme({ dimensions: 'fixed' }).space
  );
  expect(createTheme({}).sizes).toEqual(
    createTheme({ dimensions: 'fixed' }).sizes
  );
});

test('allows to use fluid dimensions', () => {
  expect(createTheme({ dimensions: 'fluid' }).space).toMatchInlineSnapshot(`
    {
      "large-1": "clamp(2rem, 4vw, 3rem)",
      "large-2": "clamp(4rem, 5vw, 5rem)",
      "large-3": "clamp(5rem, 7vw, 7.5rem)",
      "medium-1": "clamp(1rem, 2vw, 1.5rem)",
      "medium-2": "clamp(1.5rem, 3vw, 2rem)",
      "none": 0,
      "small-1": "clamp(.5rem, 1vw, 1rem)",
    }
  `);
  expect(createTheme({ dimensions: 'fluid' }).sizes).toMatchInlineSnapshot(`
    {
      "large-1": "clamp(10rem, 20vw, 15rem)",
      "large-2": "clamp(15rem, 30vw, 20rem)",
      "large-3": "clamp(20rem, 40vw, 30rem)",
      "medium-1": "clamp(4rem, 5vw, 5rem)",
      "medium-2": "clamp(5rem, 7vw, 7.5rem)",
      "medium-3": "clamp(7.5rem, 10vw, 10rem)",
      "none": 0,
      "small-1": "clamp(1rem, 2vw, 1.5rem)",
      "small-2": "clamp(1.5rem, 3vw, 2rem)",
      "small-3": "clamp(2rem, 4vw, 3rem)",
    }
  `);
});

test('allows to add component styles', () => {
  const { components } = createTheme({
    colors: ['brand', 'red'],
    components: {
      Badge: {
        base: {
          color: 'brand.primary',
        },
      },
    },
  });

  expect({ components }).toMatchInlineSnapshot(`
    {
      "components": {
        "Badge": {
          "base": {
            "color": "brand.primary",
          },
        },
      },
    }
  `);
});
