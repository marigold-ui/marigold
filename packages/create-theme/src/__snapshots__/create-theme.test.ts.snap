// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`created theme comes with defaults for "borderWidths" 1`] = `
{
  "large-1": "5px",
  "large-2": "10px",
  "large-3": "25px",
  "medium-1": "2px",
  "none": 0,
  "small-1": "1px",
}
`;

exports[`created theme comes with defaults for "breakpoints" 1`] = `
[
  "40em",
  "52em",
  "64em",
]
`;

exports[`created theme comes with defaults for "fontWeights" 1`] = `
{
  "bold": 700,
  "heavy": 900,
  "light": 300,
  "medium": 500,
  "regular": 400,
  "thin": 200,
}
`;

exports[`created theme comes with defaults for "letterSpacings" 1`] = `
{
  "large-1": ".750em",
  "large-2": "1em",
  "medium-1": ".075em",
  "medium-2": ".150em",
  "medium-3": ".500em",
  "none": "normal",
  "small-1": "-.05em",
  "small-2": ".025em",
  "small-3": ".050em",
}
`;

exports[`created theme comes with defaults for "lineHeights" 1`] = `
{
  "large-1": 1.75,
  "large-2": 2,
  "medium-1": 1.375,
  "medium-2": 1.5,
  "small-1": 0.95,
  "small-2": 1.1,
  "small-3": 1.25,
}
`;

exports[`created theme comes with defaults for "opacities" 1`] = `
{
  "dimmed": 0.5,
  "faded": 0.75,
  "ghost": 0.25,
  "solid": 1,
  "transparent": 0,
}
`;

exports[`created theme comes with defaults for "radii" 1`] = `
{
  "large-1": "4rem",
  "large-2": "8rem",
  "medium-1": "1rem",
  "medium-2": "2rem",
  "none": 0,
  "round": "1e5px",
  "small-1": "2px",
  "small-2": "5px",
}
`;

exports[`created theme comes with defaults for "zIndices" 1`] = `
{
  "1": 1,
  "2": 2,
  "3": 3,
  "4": 4,
  "5": 5,
  "important": 2147483647,
}
`;
